Fail the query when IN or NOT IN filter contains NULL.

shenyu0127 opened the pull request.

ankitsultana commented:
@shenyu0127 : can you share what was the previous behavior was for these queries?

shenyu0127 commented:
> @shenyu0127 : can you share what was the previous behavior was for these queries?

The previous behavior was to throw a different error from segment pruning: 
Currently the query `SELECT * FROM airlineStats WHERE AirTime IN (1, 2, null)` gets the following error
```
[
  {
    "errorCode": 200,
    "message": "QueryExecutionError:\norg.apache.pinot.spi.exception.BadQueryRequestException: java.lang.IllegalArgumentException: Cannot convert value: 'null' to type: INT\n\tat org.apache.pinot.core.query.pruner.ValueBasedSegmentPruner.convertValue(ValueBasedSegmentPruner.java:157)\n\tat org.apache.pinot.core.query.pruner.ValueBasedSegmentPruner$ValueCache$CachedValue.ensureDataType(ValueBasedSegmentPruner.java:223)\n\tat org.apache.pinot.core.query.pruner.ValueBasedSegmentPruner$ValueCache.get(ValueBasedSegmentPruner.java:197)\n\tat org.apache.pinot.core.query.pruner.ColumnValueSegmentPruner.pruneInPredicate(ColumnValueSegmentPruner.java:147)\n...\nCaused by: java.lang.IllegalArgumentException: Cannot convert value: 'null' to type: INT\n\tat org.apache.pinot.spi.data.FieldSpec$DataType.convertInternal(FieldSpec.java:565)\n\tat org.apache.pinot.core.query.pruner.ValueBasedSegmentPruner.convertValue(ValueBasedSegmentPruner.java:155)\n\t... 22 more"
  }
]
```

Jackie-Jiang reviewed (commented):
Can you also check if we have proper behavior on `= NULL`, `!= NULL`, `> NULL` etc? Probably we should avoid explicit `NULL` in all the predicate?

Jackie-Jiang commented:
Please also add some description to the PR and some comment in the code explaining why we don't allow NULL in IN/NOT IN

shenyu0127 commented:
> Can you also check if we have proper behavior on `= NULL`, `!= NULL`, `> NULL` etc? Probably we should avoid explicit `NULL` in all the predicate?

Good point. Changed the PR to avoid explicit `NULL` in all the predicates.

> Please also add some description to the PR and some comment in the code explaining why we don't allow NULL in IN/NOT IN

Added a function comment.

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:
```diff
@@ -267,6 +270,35 @@ private static void validateDistinctQuery(PinotQuery pinotQuery)
+   *   - For most of the filters (e.g. GREATER_THAN, LIKE), the rhs being NULL leads to no record matched.
+   *   - For IN, adding NULL to the rhs list does not change the matched records.
+   *   - For NOT IN, adding NULL to the rhs list leads to no record matched.
+   *   - For "= NULL" and "!= NULL", we have the IS_NULL and IS_NOT_NULL as workarounds.
```
Semantically `= NULL` and `!= NULL` is not the same as `IS_NULL` and `IS_NOT_NULL`. They will always return `null` which is not really useful

shenyu0127 replied:
You are right. Corrected the function comment.

Jackie-Jiang reviewed (commented):
Please add some description to the PR

shenyu0127 reviewed (commented):
> Please add some description to the PR

Done.
